### ISSUE TITLE
Speed up kuramato() function

### ISSUE THREAD

**Issue by vagechirkov** [neurolib-dev/neurolib#157]
Dear all,

I am planning to use `neurolib.utils.functions.kuromato` function to estimate Kuramoto order parameter but I faced the problem that running this function takes too long. 

Here is the time profiler report (it takes almost 10min to calculate Kuramoto order parameter for timeseries of 8 nodes and 20 seconds):

<details>

```bash

Timer unit: 1e-06 s

Total time: 588.338 s
File: /opt/anaconda3/envs/neurolib/lib/python3.7/site-packages/neurolib/utils/functions.py
Function: kuramoto at line 9

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     9                                           def kuramoto(traces, dt=0.1, smoothing=0.0, peakrange=[0.1, 0.2]):
    10                                               """
    11                                               Computes the Kuramoto order parameter of a timeseries which is a measure for synchrony.
    12                                               Can smooth timeseries if there is noise.
    13                                               Peaks are then detected using a peakfinder. From these peaks a phase is derived and then
    14                                               the amount of phase synchrony (the Kuramoto order parameter) is computed.
    15                                           
    16                                               :param traces: Multidimensional timeseries array
    17                                               :type traces: numpy.ndarray
    18                                               :param dt: Integration time step
    19                                               :type dt: float
    20                                               :param smoothing: Gaussian smoothing strength
    21                                               :type smoothing: float, optional
    22                                               :param peakrange: Width range of peaks for peak detection with `scipy.signal.find_peaks_cwt`
    23                                               :type peakrange: list[float], length 2
    24                                           
    25                                               :return: Timeseries of Kuramoto order paramter
    26                                               :rtype: numpy.ndarray
    27                                               """
    28         1         39.0     39.0      0.0      phases = []
    29         1          3.0      3.0      0.0      nTraces = len(traces)
    30         9         12.0      1.3      0.0      for n in range(nTraces):
    31         8     209572.0  26196.5      0.0          tList = np.dot(range(len(traces[n])), dt / 1000)
    32         8        102.0     12.8      0.0          a = traces[n]
    33                                           
    34                                                   # find peaks
    35         8         27.0      3.4      0.0          if smoothing > 0:
    36                                                       a = scipy.ndimage.filters.gaussian_filter(traces[n], smoothing)  # smooth data
    37         8  579281716.0 72410214.5     98.5          maximalist = scipy.signal.find_peaks_cwt(a, np.arange(peakrange[0], peakrange[1]))
    38         8        243.0     30.4      0.0          maximalist = np.append(maximalist, len(traces[n]) - 1).astype(int)
    39                                           
    40         8          9.0      1.1      0.0          if len(maximalist) > 1:
    41         8         12.0      1.5      0.0              phases.append([])
    42         8          6.0      0.8      0.0              lastMax = 0
    43      4384       3695.0      0.8      0.0              for m in maximalist:
    44   1604368    1019132.0      0.6      0.2                  for t in range(lastMax, m):
    45                                                               # compute instantaneous phase
    46   1599992    1871774.0      1.2      0.3                      phi = 2 * np.pi * float(t - lastMax) / float(m - lastMax)
    47   1599992    1174814.0      0.7      0.2                      phases[n].append(phi)
    48      4376       2871.0      0.7      0.0                  lastMax = m
    49         8          6.0      0.8      0.0              phases[n].append(2 * np.pi)
    50                                                   else:
    51                                                       logging.warning("Kuramoto: No peaks found, returning 0.")
    52                                                       return 0
    53                                           
    54                                               # determine kuramoto order paramter
    55         1          1.0      1.0      0.0      kuramoto = []
    56    200001     139197.0      0.7      0.0      for t in range(len(tList)):
    57    200000     130176.0      0.7      0.0          R = 1j * 0
    58   1800000    1258227.0      0.7      0.2          for n in range(nTraces):
    59   1600000    2636652.0      1.6      0.4              R += np.exp(1j * phases[n][t])
    60    200000     184074.0      0.9      0.0          R /= nTraces
    61    200000     425820.0      2.1      0.1          kuramoto.append(np.absolute(R))
    62                                           
    63         1          1.0      1.0      0.0      return kuramoto


```

</details>

Here is the full code to reproduce the line profiler output:

<details>

```python

from neurolib.utils.loadData import Dataset
from neurolib.models.aln import ALNModel
from neurolib.utils.functions import kuramoto
from line_profiler import LineProfiler

ds = Dataset("gw")
model = ALNModel(Cmat=ds.Cmat, Dmat=ds.Dmat)
model.params['dt'] = 0.1
model.params['duration'] = 20 * 1000  # ms

# add custom parameter for downsampling results
# 10 ms sampling steps for saving data, should be multiple of dt
model.params['save_dt'] = 10.0
model.params["tauA"] = 600.0
model.params["sigma_ou"] = 0.0
model.params["b"] = 20.0

model.params["Ke_gl"] = 300.0
model.params["signalV"] = 80.0
model.params["mue_ext_mean"] = 1.5
model.params["mui_ext_mean"] = 0.2

model.run()

lp = LineProfiler()
lp_wrapper = lp(kuramoto)
kur = lp_wrapper(model.rates_exc[::10])
lp.print_stats()

```

</details>

I am wondering whether using `scipy.signal.find_peaks_cwt` is essential. With the basic scipy function `scipy.signal.find_peaks` it works 63 times faster:

<details>

```bash

Timer unit: 1e-06 s

Total time: 9.29577 s
File: <ipython-input-11-6b8850e688bf>
Function: fast_kuramoto at line 2

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     2                                           def fast_kuramoto(traces, dt=0.1, smoothing=0.0):
     3                                               """
     4                                               Computes the Kuramoto order parameter of a timeseries which is a measure for synchrony.
     5                                               Can smooth timeseries if there is noise.
     6                                               Peaks are then detected using a peakfinder. From these peaks a phase is derived and then
     7                                               the amount of phase synchrony (the Kuramoto order parameter) is computed.
     8                                           
     9                                               :param traces: Multidimensional timeseries array
    10                                               :type traces: numpy.ndarray
    11                                               :param dt: Integration time step
    12                                               :type dt: float
    13                                               :param smoothing: Gaussian smoothing strength
    14                                               :type smoothing: float, optional
    15                                               :param peakrange: Width range of peaks for peak detection with `scipy.signal.find_peaks_cwt`
    16                                               :type peakrange: list[float], length 2
    17                                           
    18                                               :return: Timeseries of Kuramoto order paramter
    19                                               :rtype: numpy.ndarray
    20                                               """
    21         1          3.0      3.0      0.0      phases = []
    22         1          2.0      2.0      0.0      nTraces = len(traces)
    23         9         12.0      1.3      0.0      for n in range(nTraces):
    24         8     189800.0  23725.0      2.0          tList = np.dot(range(len(traces[n])), dt / 1000)
    25         8         99.0     12.4      0.0          a = traces[n]
    26                                           
    27                                                   # find peaks
    28         8         27.0      3.4      0.0          if smoothing > 0:
    29                                                       a = scipy.ndimage.filters.gaussian_filter(traces[n], smoothing)  # smooth data
    30         8      27252.0   3406.5      0.3          maximalist = scipy.signal.find_peaks(a, distance=10, prominence=5)[0]
    31         8        239.0     29.9      0.0          maximalist = np.append(maximalist, len(traces[n]) - 1).astype(int)
    32                                           
    33         8          8.0      1.0      0.0          if len(maximalist) > 1:
    34         8         11.0      1.4      0.0              phases.append([])
    35         8          6.0      0.8      0.0              lastMax = 0
    36      4345       3584.0      0.8      0.0              for m in maximalist:
    37   1604329    1052751.0      0.7     11.3                  for t in range(lastMax, m):
    38                                                               # compute instantaneous phase
    39   1599992    1912686.0      1.2     20.6                      phi = 2 * np.pi * float(t - lastMax) / float(m - lastMax)
    40   1599992    1197383.0      0.7     12.9                      phases[n].append(phi)
    41      4337       3018.0      0.7      0.0                  lastMax = m
    42         8          7.0      0.9      0.0              phases[n].append(2 * np.pi)
    43                                                   else:
    44                                                       logging.warning("Kuramoto: No peaks found, returning 0.")
    45                                                       return 0
    46                                           
    47                                               # determine kuramoto order paramter
    48         1          1.0      1.0      0.0      kuramoto = []
    49    200001     143749.0      0.7      1.5      for t in range(len(tList)):
    50    200000     134629.0      0.7      1.4          R = 1j * 0
    51   1800000    1307135.0      0.7     14.1          for n in range(nTraces):
    52   1600000    2686423.0      1.7     28.9              R += np.exp(1j * phases[n][t])
    53    200000     190310.0      1.0      2.0          R /= nTraces
    54    200000     446639.0      2.2      4.8          kuramoto.append(np.absolute(R))
    55                                           
    56         1          1.0      1.0      0.0      return kuramoto


```

</details>

The correlation between the results of the `fast_kuramoto` and `kuramoto` is around 0.99. So the output is almost identical.

Actually, 9 seconds is also too much, so I used `numba` to speed up the code a bit more. Here is the final version:

<details>

```python

from numba import jit

def fast_numba_kuramoto(traces, smoothing=5.0, dt=0.1):
    nTraces, nTimes = traces.shape
    phases = np.empty_like(traces)
    for n in range(nTraces):
        a = traces[n]
        # find peaks
        if smoothing > 0:
            # smooth data
            a = scipy.ndimage.filters.gaussian_filter(traces[n], smoothing)
        maximalist = scipy.signal.find_peaks(a, distance=10, prominence=5)[0]
        maximalist = np.append(maximalist, len(traces[n])-1).astype(int)

        if len(maximalist) > 1:
            phases[n, :] = _estimate_phase(maximalist, nTimes)
        else:
            return 0

    # determine kuramoto order paramter
    kuramoto = _estimate_r(nTraces, nTimes, phases)
    return kuramoto


@jit(nopython=True)
def _estimate_phase(maximalist, n_times):
    lastMax = 0
    phases = np.empty((n_times), dtype=np.float64)
    n = 0
    for m in maximalist:
        for t in range(lastMax, m):
            phi = 2 * np.pi * float(t - lastMax) / float(m - lastMax)
            phases[n] = phi
            n += 1
        lastMax = m
    phases[-1] = 2 * np.pi
    return phases


@jit(nopython=True)
def _estimate_r(ntraces, times, phases):
    kuramoto = np.empty((times), dtype=np.float64)
    for t in range(times):
        R = 1j*0
        for n in range(ntraces):
            R += np.exp(1j * phases[n, t])
        R /= ntraces
        kuramoto[t] = np.absolute(R)
    return kuramoto


```

</details>

The final version takes around 63ms.

Looking forward to your comments.

Valerii

**Comment by caglorithm** [comment]
Hi Valierii 

That's an amazing issue to open :)

> I am wondering whether using scipy.signal.find_peaks_cwt is essential. With the basic scipy function scipy.signal.find_peaks it works 63 times faster

Wow, I never realized the vast difference. `scipy.signal.find_peaks` is definitely the way to go then, if this is the case!

> 9 seconds is also too much, so I used numba to speed up the code a bit more. 

Ok, it's only getting better! This is great news, I suggest we update the `kuramoto` itself, instead of using a separate `fast_ kuramoto`. What do you think? If you agree, feel free to fork the `master` branch and submit a PR of the new function. I'd be happy to merge it in.

Have a great day!

**Comment by jajcayn** [comment]
Resolved in PR #158
